File: web/src/lib/api/thirteenf.ts
/**
 * 13F institutional holdings API client helpers.
 */
import { apiFetch } from "./client"

// ---------------------------------------------------------------------------
// Types matching Python Pydantic schemas (api/src/margin_api/schemas/thirteenf.py)
// ---------------------------------------------------------------------------

export interface HolderResponse {
  manager_name: string
  tier: string
  shares_held: number
  value_millions: number
  shares_changed: number
  pct_portfolio: number | null
  is_new_position: boolean
  quarters_held: number | null
}

export interface HoldingsSummary {
  total_holders: number
  curated_holders: number
  net_shares_changed: number
  signal_score: number
}

export interface HoldingsResponse {
  ticker: string
  period_of_report: string
  curated_holders: HolderResponse[]
  other_holders: HolderResponse[]
  summary: HoldingsSummary
}

export interface HoldingsHistoryQuarter {
  period: string
  curated_holders: number
  total_holders: number
  total_shares: number
  net_change: number
}

export interface HoldingsHistoryResponse {
  ticker: string
  quarters: HoldingsHistoryQuarter[]
}

export interface ManagerResponse {
  id: number
  name: string
  tier: string
  aum_millions: number | null
  total_holdings: number
  top_positions: string[]
  last_filing: string | null
  period_of_report: string | null
}

export interface PortfolioHolding {
  ticker: string | null
  cusip: string
  shares_held: number
  value_millions: number
  pct_portfolio: number
  shares_changed: number
  is_new_position: boolean
}

export interface ChangesSummary {
  new_positions: string[]
  exited_positions: string[]
  increased: number
  decreased: number
  unchanged: number
}

export interface ManagerPortfolioResponse {
  manager: string
  period_of_report: string
  aum_millions: number | null
  holdings: PortfolioHolding[]
  changes_summary: ChangesSummary
}

export interface OverlapEntry {
  ticker: string
  holder_count: number
  curated_count: number
}

export interface CrowdedTrade {
  ticker: string
  holder_count: number
  concentration_pct: number
  total_value_millions: number
}

export interface OverlapResponse {
  period_of_report: string
  most_held: OverlapEntry[]
  crowded_trades: CrowdedTrade[]
  total_managers: number | null
}

export interface NewPositionEntry {
  ticker: string
  managers: string[]
  total_new_funds: number
  curated_new_funds: number
  total_value_millions: number
}

export interface NewPositionResponse {
  period_of_report: string
  previous_quarter: string
  new_positions: NewPositionEntry[]
}

export interface ClonePosition {
  ticker: string
  target_weight: number
}

export interface ClonePerformance {
  return_1y: number | null
  cagr_3y: number | null
  max_drawdown: number | null
  sharpe: number | null
}

export interface CloneResponse {
  manager: string
  strategy: string
  period_of_report: string
  positions: ClonePosition[]
  historical_performance: ClonePerformance | null
}

// Market Pulse types

export interface SectorFlowItem {
  sector: string
  net_shares: number
  direction: "up" | "down" | "flat"
}

export interface ConsensusPick {
  ticker: string
  curated_holders: number
  agreement_pct: number
}

export interface MarketPulseResponse {
  breadth_pct: number
  breadth_direction: "up" | "down" | "flat"
  sector_flows: SectorFlowItem[]
  consensus_picks: ConsensusPick[]
  flow_trend_pct: number
  flow_trend_direction: "up" | "down" | "flat"
  as_of_quarter: string
}

// ---------------------------------------------------------------------------
// API functions
// ---------------------------------------------------------------------------

export function getHoldings(ticker: string): Promise<HoldingsResponse> {
  return apiFetch<HoldingsResponse>(`/api/v1/13f/holdings/${ticker.toUpperCase()}`)
}

export function getHoldingsHistory(ticker: string, limit = 10): Promise<HoldingsHistoryResponse> {
  return apiFetch<HoldingsHistoryResponse>(
    `/api/v1/13f/holdings/${ticker.toUpperCase()}/history?limit=${limit}`,
  )
}

export function getManagers(tier?: string): Promise<ManagerResponse[]> {
  const params = tier ? `?tier=${tier}` : ""
  return apiFetch<ManagerResponse[]>(`/api/v1/13f/managers${params}`)
}

export function getManagerPortfolio(
  managerId: number,
  period?: string,
): Promise<ManagerPortfolioResponse> {
  const params = period ? `?period=${period}` : ""
  return apiFetch<ManagerPortfolioResponse>(
    `/api/v1/13f/managers/${managerId}/portfolio${params}`,
  )
}

export function getOverlap(): Promise<OverlapResponse> {
  return apiFetch<OverlapResponse>("/api/v1/13f/analytics/overlap")
}

export function getNewPositions(): Promise<NewPositionResponse> {
  return apiFetch<NewPositionResponse>("/api/v1/13f/analytics/new-positions")
}

export function getClonePortfolio(
  managerId: number,
  strategy = "equal_weight_top_20",
): Promise<CloneResponse> {
  return apiFetch<CloneResponse>(
    `/api/v1/13f/analytics/clone/${managerId}?strategy=${strategy}`,
  )
}

export async function getMarketPulse(): Promise<MarketPulseResponse> {
  return apiFetch<MarketPulseResponse>("/api/v1/13f/analytics/market-pulse")
}

File: api/src/margin_api/schemas/thirteenf.py
"""13F holdings API response schemas."""

from __future__ import annotations

from datetime import date

from pydantic import BaseModel


class HolderResponse(BaseModel):
    manager_name: str
    tier: str
    shares_held: int
    value_millions: float
    shares_changed: int
    pct_portfolio: float | None = None
    is_new_position: bool = False
    quarters_held: int | None = None


class HoldingsSummary(BaseModel):
    total_holders: int
    curated_holders: int
    net_shares_changed: int
    signal_score: float


class HoldingsResponse(BaseModel):
    ticker: str
    period_of_report: date
    curated_holders: list[HolderResponse]
    other_holders: list[HolderResponse]
    summary: HoldingsSummary


class HoldingsHistoryQuarter(BaseModel):
    period: str
    curated_holders: int
    total_holders: int
    total_shares: int
    net_change: int


class HoldingsHistoryResponse(BaseModel):
    ticker: str
    quarters: list[HoldingsHistoryQuarter]


class ManagerResponse(BaseModel):
    id: int
    name: str
    tier: str
    aum_millions: float | None = None
    total_holdings: int
    top_positions: list[str]
    last_filing: date | None = None
    period_of_report: date | None = None


class PortfolioHolding(BaseModel):
    ticker: str | None = None
    cusip: str
    shares_held: int
    value_millions: float
    pct_portfolio: float
    shares_changed: int
    is_new_position: bool = False


class ChangesSummary(BaseModel):
    new_positions: list[str]
    exited_positions: list[str]
    increased: int
    decreased: int
    unchanged: int


class ManagerPortfolioResponse(BaseModel):
    manager: str
    period_of_report: date
    aum_millions: float | None = None
    holdings: list[PortfolioHolding]
    changes_summary: ChangesSummary


class OverlapEntry(BaseModel):
    ticker: str
    holder_count: int
    curated_count: int


class CrowdedTrade(BaseModel):
    ticker: str
    holder_count: int
    concentration_pct: float
    total_value_millions: float


class OverlapResponse(BaseModel):
    period_of_report: date
    most_held: list[OverlapEntry]
    crowded_trades: list[CrowdedTrade]
    total_managers: int | None = None


class NewPositionEntry(BaseModel):
    ticker: str
    managers: list[str]
    total_new_funds: int
    curated_new_funds: int
    total_value_millions: float


class NewPositionResponse(BaseModel):
    period_of_report: date
    previous_quarter: date
    new_positions: list[NewPositionEntry]


class ClonePerformance(BaseModel):
    return_1y: float | None = None
    cagr_3y: float | None = None
    max_drawdown: float | None = None
    sharpe: float | None = None


class ClonePosition(BaseModel):
    ticker: str
    target_weight: float


class CloneResponse(BaseModel):
    manager: str
    strategy: str
    period_of_report: date
    positions: list[ClonePosition]
    historical_performance: ClonePerformance | None = None


class SectorFlowItem(BaseModel):
    sector: str
    net_shares: int
    direction: str  # "up" | "down" | "flat"


class ConsensusPick(BaseModel):
    ticker: str
    curated_holders: int
    agreement_pct: float


class MarketPulseResponse(BaseModel):
    breadth_pct: float
    breadth_direction: str  # "up" | "down" | "flat"
    sector_flows: list[SectorFlowItem]
    consensus_picks: list[ConsensusPick]
    flow_trend_pct: float
    flow_trend_direction: str  # "up" | "down" | "flat"
    as_of_quarter: str
